File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        res = 0
        index_map = {}  # Dictionary to store the last index of each character

        for r in range(len(s)):
            if s[r] in index_map and index_map[s[r]] >= l:
                # Move `l` to the right of the last occurrence of s[r]
                l = index_map[s[r]] + 1
            
            # Update the most recent index of s[r]
            index_map[s[r]] = r

            # Update the result with the length of the current window
            res = max(res, r - l + 1)
        
        return res

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int right = 0;
        int res = 0;
        
        if (s.length() == 0) return 0;
        
        HashSet<Character> visited = new HashSet<Character>();

        while (right < s.length()) {
            if (!visited.contains(s.charAt(right))) {
                visited.add(s.charAt(right));
                right++;
                res = Math.max(res, right - left);
            } else {
                visited.remove(Character.valueOf(s.charAt(left)));
                left++;
            }
        } 

        return res;
    }
}